src/os/bluestore/BlueStore.h

```diff
@@ -2937,7 +2941,7 @@ class BlueStore : public ObjectStore,
 private:
   void _deferred_submit_unlock(OpSequencer *osr);
   void _deferred_aio_finish(OpSequencer *osr);
-  int _deferred_replay();
+  int _deferred_replay(std::vector<std::string>* keys_to_remove);
```

Comment:
You might want to pass a pointer to target Transaction instead. And avoid later copying from vector to that transaction.
I'm a bit scary of potential maximum size of such a transaction though, which is applicable to vector approach as well.

Reply:
The size of keys_to_remove is limited by amount of deferred writes that we created but not deleted yet.
It cannot be too many since BlueStore keeps in-memory block data related to deferred transactions.

Transaction is good idea, but cannot fly.
I started with it, but it turned out that db->get_transaction() is tied to DB instance.
If I fill this transaction in read-only DB, but then applying to to read-write DB causes assert.